homeassistant/components/dhcp/__init__.py

```diff
@@ -218,22 +229,30 @@ def async_process_client(
                 if entry := self.hass.config_entries.async_get_entry(entry_id):
                     device_domains.add(entry.domain)
 
-        for entry in self._integration_matchers:
-            if entry.get(REGISTERED_DEVICES) and not entry["domain"] in device_domains:
+        for matcher in self._integration_matchers:
+            if (
+                matcher.get(REGISTERED_DEVICES)
+                and not matcher["domain"] in device_domains
+            ):
                 continue
 
-            if MAC_ADDRESS in entry and not fnmatch.fnmatch(
-                uppercase_mac, entry[MAC_ADDRESS]
+            if (
+                (matcher_mac := matcher.get(MAC_ADDRESS)) is not None
+                and isinstance(matcher_mac, str)
```

Comment:
Don't we know that the mac address is a string?

Reply:
Changed to an assert in 556ddb36bc